### ISSUE TITLE
Web/HTML5 export

### ISSUE THREAD

**Issue by ashtonmeuser** [ashtonmeuser/godot-wasm#18]
Just allowing Godot Wasm to be used as a module is not enough to export for web/HTML5. This is because Wasmer (the Wasm runtime used by this project) is not able to be comiled to Wasm by Emscripten which is what is happening internally when compiling the web/HTML5 Godot export templates. The following error is presented.

```
[bin/godot.javascript.opt.js] Implicit dependency `modules/wasm/libwasmer.javascript.opt.a' not found, needed by target `bin/godot.javascript.opt.js'.
```

I see a few solutions to this but none that are a simple addition to this project.
1. Use a Wasm runtime that is able to be compiled to Wasm such as [Wasm3](https://github.com/wasm3/wasm3). This library is should be able to be compiled by Emscripten. Unfourtunately, it does not seem to adhere to the [Wasm C API](https://github.com/WebAssembly/wasm-c-api) and would therefore require a significant rework of this project or a shim to add API equivalence.
2. Create a small shim library that exposes the [Wasm C API](https://github.com/WebAssembly/wasm-c-api) (which is used by this project) and translates the relevant Wasm interactions to JavaScript. Because the browser has a Wasm runtime, including another runtime e.g. Wasmer or Wasm3 seems superfluous. This seems like the more sensible option.

**Comment by fire** [comment]
In https://github.com/ashtonmeuser/godot-wasm/blob/master/config.py I would return False if the platform is web.

**Comment by ashtonmeuser** [comment]
> In https://github.com/ashtonmeuser/godot-wasm/blob/master/config.py I would return False if the platform is web for now.

Done. Only supporting Linux, macOS, and Windows.

**Comment by ShalokShalom** [comment]
Sorry, I have to ask this probably naive question: Why not using a runtime, that can be compiled?

**Comment by ashtonmeuser** [comment]
> Sorry, I have to ask this probably naive question: Why not using a runtime, that can be compiled?

I think my comment at the top of this briefly touches on the issue but I'll expand here. 

The Godot Wasm project relies heavily on the [Wasm C API](https://github.com/WebAssembly/wasm-c-api), largely because this is the closest thing we've got to a standardized API. This allows swapping the Wasm runtime with relatively low effort (see [documentation](https://github.com/ashtonmeuser/godot-wasm/wiki/Development#runtime-options)).

In order to run on the web, Godot compiles the your project to Wasm using Emscripten. To use Wasm via an included Wasm runtime (as Godot Wasm does), and without using a JS bridge as mentioned above, the runtime itself would need to be compiled to Wasm. Neither Wasmer nor Wasmtime, Godot Wasm's supported runtimes, support this. 

In short, we'd need a Wasm runtime that supports _both_ 1) compiling to Wasm, and 2) the Wasm C API. I have not seen a runtime meeting both criteria.

**Comment by tjpalmer** [comment]
Shim to builtin browser wasm support makes most sense, anyway, since it's likely to be way more performant than a wasm interpreter running in wasm.

**Comment by ShalokShalom** [comment]
You could use WAMR (WebAssembly Micro Runtime) with the interpreter (has an AoT as well, that won't compile to wasm) 

That's relatively slow, though

**Comment by ashtonmeuser** [comment]
> Using WAMR seems to be viable.

Wow, incredible work, @basicer! This project has been slightly deprioritized as of late but if you can give a little more information as to how you implemented this, I'll include WAMR as an alternative runtime (in addition to Wasmer and Wasmtime).

**Comment by basicer** [comment]
Thanks! There's quite a bit of duct tape holding this all up, but I'd like to get the various changed needed contributed upstream soon.

Planning on switching to this from my current solution (https://github.com/godotengine/godot/pull/95788) if performance is comparable and it works on all current godot platforms. Both projects have benchmarks claiming to be faster than the other, but wamr seems to be more activity maintained though a bit less portable and a bit more "bloated" in terms of amount of code.

**Comment by ashtonmeuser** [comment]
Began looking into this. This should be a small lift although is complicated by WAMR not including libraries in release artifacts (unlike Wasmtime and Wasmer). See https://github.com/bytecodealliance/wasm-micro-runtime/issues/4602.

**Comment by basicer** [comment]
My solution for this is that I wrote a python script that extracts the important bits from WARM for the fast interpreter similar to what the godot project does to manage their dependences.  That way I don't have to worry about upstream producing build artifacts for every language. As a module the godot build scripts can then just build it easily for every supported configuration.
